spyder/api/folding.py

```diff
@@ -0,0 +1,365 @@
+        # the previous blank line will keep the trigger state and the new line
+        # (which actually contains the trigger) must use the prev state (
+        # and prev state must then be reset).
+        prev = current_block.previous()  # real prev block (may be blank)
+        if (prev and prev.isValid() and prev.text().strip() == '' and
+                TextBlockHelper.is_fold_trigger(prev)):
+            # prev line has the correct trigger fold state
+            TextBlockHelper.set_collapsed(
+                current_block, TextBlockHelper.is_collapsed(
+                    prev))
+            # make empty line not a trigger
+            TextBlockHelper.set_fold_trigger(prev, False)
+            TextBlockHelper.set_collapsed(prev, False)
+
+    def detect_fold_level(self, prev_block, block):
+        """
+        Detects the block fold level.
+
+        The default implementation is based on the block **indentation**.
+
+        .. note:: Blocks fold level must be contiguous, there cannot be
+            a difference greater than 1 between two successive block fold
+            levels.
+
+        :param prev_block: first previous **non-blank** block or None if this
+            is the first line of the document
+        :param block: The block to process.
+        :return: Fold level
+        """
+        raise NotImplementedError
+
+
+class IndentFoldDetector(FoldDetector):
+    """Simple fold detector based on the line indentation level."""
+
+    def detect_fold_level(self, prev_block, block):
+        """
+        Detects fold level by looking at the block indentation.
+
+        :param prev_block: previous text block
+        :param block: current block to highlight
+        """
+        text = block.text()
+        # round down to previous indentation guide to ensure contiguous block
+        # fold level evolution.
+        return (len(text) - len(text.lstrip())) // len(self.editor.indent_chars)
+
+
+class CharBasedFoldDetector(FoldDetector):
+    """
+    Fold detector based on trigger charachters (e.g. a { increase fold level
+    and } decrease fold level).
+    """
+    def __init__(self, open_chars=('{'), close_chars=('}')):
+        super(CharBasedFoldDetector, self).__init__()
+        self.open_chars = open_chars
+        self.close_chars = close_chars
+
+    def detect_fold_level(self, prev_block, block):
+        if prev_block:
+            prev_text = prev_block.text().strip()
+        else:
+            prev_text = ''
+        text = block.text().strip()
+        if text in self.open_chars:
+            return TextBlockHelper.get_fold_lvl(prev_block) + 1
+        if prev_text.endswith(self.open_chars) and prev_text not in \
+                self.open_chars:
+            return TextBlockHelper.get_fold_lvl(prev_block) + 1
+        if self.close_chars in prev_text:
+            return TextBlockHelper.get_fold_lvl(prev_block) - 1
+        return TextBlockHelper.get_fold_lvl(prev_block)
+
+
+class FoldScope(object):
```

Comment:
When moving this `widgets/sourcecode/folding.py` please put this class at the beginning of that file.